DasharoPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c

```diff
@@ -1067,6 +1070,168 @@ WarnIfRecoveryBoot (
+  EFI_EVENT      Events[1];
+  BOOLEAN        CursorVisible;
+  UINTN          CurrentAttribute;
+  UINTN          Index;
+  EFI_INPUT_KEY  Key;
+  UINT32         TpmHashAlgorithmBitmap;
+  UINT32         ActivePcrBanks;
+  UINT32         RequestedActivePcrBanks;
+  UINTN          Size;
+  UINTN          i;
+  UINTN          OptionCount = 0;
+  UINTN          AvlIdx[ARRAY_SIZE(mHashAlgs)] = {0};
+  CHAR16         OptLine[81];
+  UINT32         SelectedMask;
+
+  Status = Tpm2GetCapabilitySupportedAndActivePcrs(&TpmHashAlgorithmBitmap, &ActivePcrBanks);
+  if (EFI_ERROR(Status)) {
+      return;
+  }
+  Size = sizeof(RequestedActivePcrBanks);
+  Status = gRT->GetVariable(
+            REQUESTED_ACTIVE_PCR_BANKS_VARIABLE_NAME,
+            &gEfiTcg2PhysicalPresenceGuid,
+            NULL,
+            &Size,
+            &RequestedActivePcrBanks
+            );
+  // 
+  // If the variable doesn't exist, there's been no request to change the
+  // PCR banks.
+  // 
+  if (Status == EFI_NOT_FOUND) {
+    return;
+  }
+  // Clear the variable after reading
+  gRT->SetVariable(
+        REQUESTED_ACTIVE_PCR_BANKS_VARIABLE_NAME,
+        &gEfiTcg2PhysicalPresenceGuid,
+        0,
+        0,
+        NULL
+      );
+  // If they're equal, the requested bank selection was successful.
+  if (RequestedActivePcrBanks == ActivePcrBanks) {
+    return;
+  }
+  // 
+  // If they're not equal, display the popup and switch to a single bank
+  // of user's choice.
+  // 
+  ASSERT_EFI_ERROR(Status);
+  CurrentAttribute = gST->ConOut->Mode->Attribute;
+  CursorVisible    = gST->ConOut->Mode->CursorVisible;
+  gST->ConOut->EnableCursor(gST->ConOut, FALSE);
+  DrainInput();
+  Events[0] = gST->ConIn->WaitForKey;
+  // 
+  // Parse obtained available PCR banks bitmap to get the names and create
+  // an option for each available bank
+  // 
+  for (i = 0; i < ARRAY_SIZE(mHashAlgs); i++) {
+    if (TpmHashAlgorithmBitmap & (1U << i)) {
+      AvlIdx[OptionCount++] = i;
+    }
+  }
+  OptLine[0] = L'\0';
+  UnicodeSPrint(OptLine, sizeof(OptLine), L"Select bank:");
+  // Append each available bank
+  for (i = 0; i < OptionCount; i++) {
+    CHAR16 token[40];
+    UnicodeSPrint(
+      token, sizeof(token),
+      L" %u) %s", (UINT32)(i + 1), mHashAlgs[AvlIdx[i]].Name
+    );
+    StrCatS(OptLine, ARRAY_SIZE(OptLine), token);
+  }
+  
+  while (1) {
+    CreateMultiStringPopUp(
+        78,
+        9,
+        L"!!! WARNING !!!",
+        L"",
+        L"Multiple PCR banks have been selected, but the current TPM supports",
+        L"only one active bank at a time.",
+        L"",
+        OptLine,
+        L"",
+        L"Press ESC to stay with the previously active bank.",
+        L""
+    );
+
+    Status = gBS->WaitForEvent(1, Events, &Index);
+    if (EFI_ERROR(Status)) {
+      continue;
+    }
+    
+    if (Index != 0) continue;
+
+    Status = gST->ConIn->ReadKeyStroke(gST->ConIn, &Key);
+    if (EFI_ERROR(Status)) {
+      continue;

```

Comment:
Continuing on error here and for `WaitForEvent` may not be the best approach.  These functions shouldn't fail randomly and if this is something permanent the code will just hang.  Breaking out of the loop seems more preferable.